core/src/main/java/org/elasticsearch/action/admin/indices/upgrade/post/TransportUpgradeAction.java

```diff
@@ -94,20 +95,40 @@ protected UpgradeResponse newResponse(UpgradeRequest request, AtomicReferenceArr
                     Integer count = successfulPrimaryShards.get(index);
                     successfulPrimaryShards.put(index, count == null ? 1 : count + 1);
                 }
-                Version version = versions.get(index);
-                if (version == null || shardUpgradeResponse.version().onOrAfter(version) == false) {
-                    versions.put(index, shardUpgradeResponse.version());
+                Tuple<Version, org.apache.lucene.util.Version> versionTuple = versions.get(index);
+                if (versionTuple == null) {
+                    versions.put(index, new Tuple<>(shardUpgradeResponse.version(), shardUpgradeResponse.luceneVersion()));
+                } else {
+                    // We already have versions for this index - let's see if we need to update them based on the current shard
+                    Version version = versionTuple.v1();
+                    org.apache.lucene.util.Version luceneVersion = versionTuple.v2();
+                    boolean changed = false;
+                    // For the metadata we are interested in the _latest_ elasticsearch version that was processing the metadata
+                    // Since we rewrite the mapping during upgrade the metadata is always rewritten by the latest version
+                    if (shardUpgradeResponse.version().after(versionTuple.v1())) {
+                        version = shardUpgradeResponse.version();
+                        changed = true;
+                    }
+                    // For the lucene version we are interested in the _oldest_ lucene version since it determines the
+                    // oldest version that we need to support
+                    if (shardUpgradeResponse.luceneVersion().onOrAfter(versionTuple.v2()) == false) {
+                        luceneVersion = shardUpgradeResponse.luceneVersion();
+                        changed = true;
+                    }
+                    if (changed) {
+                        versions.put(index, new Tuple<>(version, luceneVersion));
+                    }
```

Comment:
maybe we don't need this `changed` logic, calling `versions.put(index, new Tuple<>(version, luceneVersion));` would always be correct?